### DISCUSSION TITLE
Output lora stack from Power Lora Loader

### DISCUSSION THREAD

**Discussion by Tekaiguy** [rgthree/rgthree-comfy#324] [Ideas]
Imagine you want to use a lora stack with two different checkpoints in the same workflow, say, for creating the original image and then using it in img2img. There's not really any way to do that with the Power Lora Loader node since it can't turn any of its widgets into inputs, and you also can't give it inputs from 2 different models.

If it could output a lora stack and send it to a combiner node that takes 'model', 'clip', and 'lora stack' as input and outputs 'model' and 'clip', then a single Power Lora Loader could be used with as many KSamplers as you want.

**Comment by TanerH** [comment]
I love the Power Lora node, but yes - being able to have the lora stack output would be great!

For my own use, I would be happy if the node could output a text string of all the enabled loras (to append to my prompt before saving - just for convenience and flexibility).  I can accomplish this if it outputs the lora stack, since there are stack -> string nodes already out there.

**Reply by SebAnt333** [reply]
I would be thrilled if the node could output a text string of all the enabled loras, so I can optionally overlay this text in final image if desired. For now, I was able to use the WAS suite lora loader that also outputs the lora name as string, string three more of these, and concatenate all lora names to accomplish this. The power lora loader is far more elegant and compact.

**Reply by Symbiomatrix** [reply]
@SebAnt333 Just did that in my side version, it's very easy. In py/power_lora_loader.py:
- Add to INPUT_TYPES the variable `OUTPUT_IS_LIST = (False, False, True)` and modify variables `RETURN_TYPES = ("MODEL", "CLIP", "STRING")` , `RETURN_NAMES = ("MODEL", "CLIP", "LORA_NAMES")`
- In load_loras, create a variable, say `lnames = []`, then fill it inside the loop (say after `if lora is not None: lnames.append(value["lora"])`). This will inherently check being enabled, strength not zero and lora exists.
- Change its return to `return (model, clip, lnames)`
- Connect the LORA_NAMES output to some text node, say pygoss' show_text.

**Comment by SebAnt333** [comment]
WOW!! 🤩 Though I don’t know any python, your instructions were crystal clear and worked like a charm. I cannot thank you enough. I’m so grateful and over the moon with your modification 🙏🏼🙏🏼🙏🏼Sent from my iPhoneOn Mar 4, 2025, at 12:21 PM, Symbiomatrix ***@***.***> wrote:﻿
@SebAnt333 Just did that in my side version, it's very easy. In py/power_lora_loader.py:

Add to INPUT_TYPES the variable OUTPUT_IS_LIST = (False, False, True) and modify variables RETURN_TYPES = ("MODEL", "CLIP", "STRING") , RETURN_NAMES = ("MODEL", "CLIP", "LORA_NAMES")
In load_loras, create a variable, say lnames = [], then fill it inside the loop (say after if lora is not None: lnames.append(value["lora"])).
return (model, clip, lnames)

**Reply by Symbiomatrix** [reply]
You're welcome. Output works great, could probably just as easily pass around the lora stack itself in some form (even just the names and then make a separate dummy node to load them in a loop, would satisfy the OP's requirement).
Problem is the input - this node seems to have a custom loading function which discards anything not related to loras, so whenever you reload the new widget isn't loaded nor created. I don't even know how to make it optional so it just throws an error.

**Reply by Tekaiguy** [reply]
I no longer have that requirement for my project, but I'm glad I could help facilitate what sounds like the start of someone's career.

Anyway, even if I was able and willing to get it working, I wanted to be able to share my workflow so a hotfix wouldn't help me there. It would be cool if you wanted to teach this stuff since there's not a lot out there right now, and you seem to be able to make it accessible.

**Comment by SebAnt333** [comment]
Using Claude which I just subscribed to, I further modified it to exclude
the .safetensors suffix but add the LoRA strength rounded to 2 decimal
places. You have opened my eyes to a whole new world of modifying nodes,
because I'm more of an artist than a programmer.
if lora_name.endswith('.safetensors'):
  lora_name = lora_name[:-12]  # Remove last 12 characters ('.safetensors')
  # Concatenate strength_model to the lora_name (with 2 decimal places)
formatted_strength = f"{strength_model:.2f}".rstrip('0').rstrip('.') if '.'
in f"{strength_model:.2f}" else f"{strength_model:.2f}"
lora_name = f"{lora_name}_{formatted_strength}"

**Reply by Symbiomatrix** [reply]
Small improvement, you can remove the .safetensors suffix more dynamically using `lora_name.rsplit(".", 1)[0]` instead of fixed 12 chars and check . This will remove everything after the last period.

Also, you can use \`\`\` some code \`\`\` to format code on git: ```some code```

And yeah, tinkering these small pet peeves away with a bit of code is very satisfying.

**Comment by SebAnt333** [comment]
Thanks for the tip - I am making the change!

**Comment by mnedix** [comment]
Heya, I also love the Power Lora and I truly believe a string output of its Loras would be a great addition. I did everything as @Symbiomatrix said and I kinda made it work. My problem is two fold: 1. it outputs only one Lora name rather than all of them and 2. it just gives me the path including the name so I have to run it through a text "stripper" to get only the Lora name. 
I guess what I would like to have is the Lora name and weight in sharp brackets <> so I can concatenate the string together with the prompt and other styles (and subsequently to be added to the Image Saver and then recognized by Civitai). 
Last but not least, I have no idea of programming, I'm just reverse engineering what I can so if anyone wants to jump in, please do not assume I necessarily know what you're talking about :)

**Reply by Symbiomatrix** [reply]
The solution I provided handles multiple lora names (outputs them as a list). Perhaps whichever node you're using for display doesn't. In any case, you can output a customised long string of names instead of list if you prefer, eg using `" ".join(lnames)` - just ask your llm of choice how to do that.
[Pygoss' repo](https://github.com/pythongosssss/ComfyUI-Custom-Scripts) along with kjnodes are my bread and butter for simple operations - you can load / save text to file + compute basic math expressions + show text with pygoss, join strings (eg add the brackets) with kjnodes. The only thing I was missing is a simple cast from text to number (in my custom nodes). I have a workflow here that does that sort of string related stuff: https://github.com/Symbiomatrix/Comfyui-Sort-Files/blob/main/SampleWorkflows/SBM_Test_I2VE_SkeletonV2.json

**Comment by giminic** [comment]
Sorry to dig up this old discussion.
It's quite easy to output a "lora_stack" (basically just a 2-dimensional array of lora files and weights) object as requested by OP.
This stack can then be utilized by other nodes (i.e. "Construct Metadata" from sageutils).
This comment is also meant to document it for myself in case of the next update overwriting my adjustments.


Just need to update 3 lines and add 2 more in "py/power_lora_loader.py":

In the current version 1.0.2604070017, update lines 34 & 35 to:
`  RETURN_TYPES = ("MODEL", "CLIP", "LORA_STACK")`
`  RETURN_NAMES = ("MODEL", "CLIP", "LORA_STACK")`
  
Below `def load_loras(self, model=None, clip=None, **kwargs):` add 1 new line:
`    lora_stack = []`

Below `model, clip = LoraLoader().load_lora(model, clip, lora, strength_model, strength_clip)` add 1 new line:
`            lora_stack.append([value['lora'], strength_model, strength_clip])`

Last, adjust the line below to:
`    return (model, clip, lora_stack)`

Pay attention to indentation as this defines, which conditional statement the action belongs to.
